<div style="border: 5px solid purple; padding: 15px; margin: 5px">
<b> Reviewer's comment 2 </b>
    
Thank you for your diligence! I've left a couple of new comments with digit 2, but there's nothing that needs to be fixed.
    
    
You did a fantastic job here. You have correctly split the data, explored the distributions, trained several models, and applied the bootstrapping technique. You have successfully estimated profit and conducted risk analysis. You have also compared regions based on risk-adjusted return to make recommendations for further development. Excellent job! 
    
    
I do not have any questions, so the project can be accepted. The next sprints will cover more advanced machine learning methods, I hope you will like it. Good luck! 
    
    

</div>

<div style="border: 5px solid purple; padding: 15px; margin: 5px">
<b> Reviewer's comment</b>
    
Hi Joseph, my name is Svetlana (https://hub.tripleten.com/u/855fe797). Congratulations on submitting another project! 🎉
    

Before we start, I want to pay your attention to the color marking:
    

   
    
<div style="border: 5px solid green; padding: 15px; margin: 5px">

Great solutions and ideas that can and should be used in the future are in green comments. Some of them are: 
    
    
- You have successfully prepared the subsets. It is important to split the data correctly in order to ensure there's no intersection;    
    
    
- You have successfully explored the regions;
   
    
- Used linear regression to make predictions;
    
    
- Calculated profit and estimated risk values; 
    
    
- Wrote an excellent conclusion! A well-written conclusion shows how the project met its objectives and provides a concise and understandable summary for those who may not have been involved in the details of the project. Good job! 

</div>
    
<div style="border: 5px solid gold; padding: 15px; margin: 5px">
<b> Reviewer's comment </b>

Yellow color indicates what should be optimized. This is not necessary, but it will be great if you make changes to this project. I've left several recommendations throughout the project. 
    
</div>
<div style="border: 5px solid red; padding: 15px; margin: 5px">
<b> Reviewer's comment </b>

Issues that must be corrected to achieve accurate results are indicated in red comments. Please note that the project cannot be accepted until these issues are resolved. 
</div>        
<hr>
    
<font color='dodgerblue'>**To sum up:**</font> you did a great job here, thank you so much for submitting the project! Actually the project is almost perfect, and the updates should not take much time. If you have any questions, please feel free to ask. I will wait the project for the second review 😊 
    

<hr>
    
Please use some color other than those listed to highlight answers to my comments.
I would also ask you **not to change, move or delete my comments** to make it easier for me to navigate during the next review.
    
<hr> 
    
✍️ Here's a nice [article](https://machinelearningmastery.com/a-gentle-introduction-to-the-bootstrap-method/) about the bootstrap method. [This](https://www.youtube.com/watch?v=O_Fj4q8lgmc&list=LL&index=6) is a good video about resampling in bootstrapping.
    

</div>

## Project Overview
In this project, I work for the fictional OilyGiant mining company. The goal is to determine the most profitable region for developing a new oil well.

To do this, I will:

- Load and prepare geological data from three regions.
- Train Linear Regression models to predict oil reserves.
- Select the 200 most promising wells out of 500 based on predictions.
- Calculate potential profit using business constraints.
- Apply bootstrapping (1000 samples) to estimate average profit, 95% confidence intervals, and risk of losses.
- Recommend the optimal region based on profit and risk.
This project uses Linear Regression only, according to business requirements.
    

In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error

In [2]:
df0 = pd.read_csv('geo_data_0.csv')
df1 = pd.read_csv('geo_data_1.csv')
df2 = pd.read_csv('geo_data_2.csv')

dfs = [df0, df1, df2]

In [3]:
for i, df in enumerate(dfs):
    print(f'\nRegion {i}')
    display(df.head())
    print(df.info())
    print(df.describe())


Region 0


id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.500419       0.250143       2.502647      92.500000
std         0.871832       0.504433       3.248248      44.288691
min        -1.408605      -0.848218     -12.088328       0.000000
25%        -0.072580      -0.200881       0.287748      56.497507
50%         0.502360       0.250252       2.515969      91.849972
75%         1.073581       0.700646       4.715088     128.564089
max         2.362331       1.3437

id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        1.141296      -4.796579       2.494541      68.825000
std         8.965932       5.119872       1.703572      45.944423
min       -31.609576     -26.358598      -0.018144       0.000000
25%        -6.298551      -8.267985       1.000021      26.953261
50%         1.153055      -4.813172       2.011479      57.085625
75%         8.621015      -1.332816       3.999904     107.813044
max        29.421755      18.7340

id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
                  f0             f1             f2        product
count  100000.000000  100000.000000  100000.000000  100000.000000
mean        0.002023      -0.002081       2.495128      95.000000
std         1.732045       1.730417       3.473445      44.749921
min        -8.760004      -7.084020     -11.970335       0.000000
25%        -1.162288      -1.174820       0.130359      59.450441
50%         0.009424      -0.009482       2.484236      94.925613
75%         1.158535       1.163678       4.858794     130.595027
max         7.238262       7.8448

In [4]:
def split_features_target(df):
    X = df[['f0', 'f1', 'f2']]
    y = df['product']
    return X, y

X_y = [split_features_target(df) for df in dfs]


## Data Preparation

The datasets contain:
- id — well identifier
- f0, f1, f2 — feature values
- product — oil reserve volume (in thousand barrels)
# Procedure
- Loaded all three datasets (geo_data_0.csv, geo_data_1.csv, geo_data_2.csv).
- Checked for missing values and verified data types.
- No missing or corrupt data found, so no cleaning was needed.
- Selected f0, f1, f2 as features and product as the target variable.
This ensured the data was ready for modeling.

We use features f0, f1, and f2 to predict the target product (thousand barrels). No additional preprocessing was required because there were no missing values and all features were numeric.

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
Good! 
    
</div>
<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
Consider displaying some charts as well. It will be easier to analyze the distributions.    
</div>

In [5]:
def train_and_evaluate_region(X, y, random_state=12345):
    X_train, X_valid, y_train, y_valid = train_test_split(
        X, y, test_size=0.25, random_state=random_state
    )
    
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    y_pred = model.predict(X_valid)
    
    rmse = mean_squared_error(y_valid, y_pred, squared=False)
    mean_pred = y_pred.mean()
    
    return {
        'model': model,
        'X_valid': X_valid,
        'y_valid': y_valid,
        'y_pred': y_pred,
        'rmse': rmse,
        'mean_pred': mean_pred
    }

In [6]:
results = []

for i, (X, y) in enumerate(X_y):
    res = train_and_evaluate_region(X, y)
    results.append(res)
    print(f'\nRegion {i}')
    print(f'RMSE: {res["rmse"]:.2f}')
    print(f'Average predicted product: {res["mean_pred"]:.2f}')


Region 0
RMSE: 37.58
Average predicted product: 92.59

Region 1
RMSE: 0.89
Average predicted product: 68.73

Region 2
RMSE: 40.03
Average predicted product: 94.97


In [7]:
for i, res in enumerate(results):
    print(f'Region {i} - mean actual: {res["y_valid"].mean():.2f}, mean predicted: {res["mean_pred"]:.2f}')

Region 0 - mean actual: 92.08, mean predicted: 92.59
Region 1 - mean actual: 68.72, mean predicted: 68.73
Region 2 - mean actual: 94.88, mean predicted: 94.97


## Model Training & Evaluation

For each region, the dataset was split into 75% training and 25% validation.  
A Linear Regression model was trained and evaluated.

## Model Performance Results

| Region | RMSE | Mean Actual | Mean Predicted |
|--------|-------|--------------|----------------|
| **0** | 37.58 | 92.08 | 92.59 |
| **1** | 0.89 | 68.72 | 68.73 |
| **2** | 40.03 | 94.88 | 94.97 |

## Findings

- Region 1 has the lowest RMSE, indicating the most accurate model.  
- Regions 0 and 2 show higher RMSE values and more variability.  
- All predicted means closely match actual means, showing no major bias.  
- However, model accuracy alone does not determine profitability — financial evaluation is required next.

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    
Correct! Region 2 is the most profitable, but it has the highest model error at the same time. 

</div>

In [8]:
BUDGET = 100_000_000  # dollars
WELLS_TO_DEVELOP = 200
WELLS_TO_STUDY = 500
PRICE_PER_BARREL = 4.5  # dollars
PRICE_PER_THOUSAND_BARRELS = PRICE_PER_BARREL * 1000  # 4500

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
 
Correct! Constant values indeed should be defined in the upper case.

</div>
<div style="border: 5px solid gold; padding: 10px; margin: 5px">
<b> Reviewer's comment </b>
    
    

We should define constants in the very beginning of the project.  
    
</div>

In [9]:
min_volume_per_well = BUDGET / (PRICE_PER_THOUSAND_BARRELS * WELLS_TO_DEVELOP)
print(f'Minimum volume per selected well to break even: {min_volume_per_well:.2f} thousand barrels')

Minimum volume per selected well to break even: 111.11 thousand barrels


In [10]:
for i, (X, y) in enumerate(X_y):
    print(f'Region {i} - mean product: {y.mean():.2f} thousand barrels')

Region 0 - mean product: 92.50 thousand barrels
Region 1 - mean product: 68.83 thousand barrels
Region 2 - mean product: 95.00 thousand barrels


<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    


So the average volumes in these three regions are less than the required value. If we are to select the wells randomly, we are probably going to lose money. That's why we need machine learning and bootstrapping to estimate profit distributions. 
    
    
</div>

In [11]:
def calculate_profit(y_true, y_pred):
    """
    y_true, y_pred are pandas Series of length 500 (validation set).
    """
    # Indices of wells sorted by predicted product (descending)
    order = y_pred.sort_values(ascending=False).index
    
    # Take top N wells
    top_indices = order[:WELLS_TO_DEVELOP]
    
    # Actual product of selected wells
    selected_products = y_true.loc[top_indices]
    
    total_product = selected_products.sum()  # thousand barrels
    
    revenue = total_product * PRICE_PER_THOUSAND_BARRELS
    profit = revenue - BUDGET
    
    return profit

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>

Correct! We select the best 200 wells by predictions, but then we need to use target values to actually calculate profit because predictions may not be accurate. 

</div>

In [12]:
for i, res in enumerate(results):
    profit = calculate_profit(res['y_valid'], pd.Series(res['y_pred'], index=res['y_valid'].index))
    print(f'Region {i} - profit estimate (single sample): {profit/1_000_000:.2f} million USD')

Region 0 - profit estimate (single sample): 33.21 million USD
Region 1 - profit estimate (single sample): 24.15 million USD
Region 2 - profit estimate (single sample): 27.10 million USD


## Preparing for Profit Calculation

### Business Conditions

- Development budget: $100 million
  
- Only 200 wells can be developed out of 500 explored

- Revenue per thousand barrels: $4,500
  
- Profit formula:  
  Profit = (Sum of selected wells’ product × 4500) − 100,000,000

---

## Break-Even Volume Calculation

Break-even volume per well:

Break-even = Budget / (4500 × 200)  
Break-even = 100,000,000 / 900,000  
Break-even = **111.11 thousand barrels**

The minimum production volume per selected well needed to avoid losses is:

**111.11 thousand barrels**

---

## Mean Product Per Region

| Region | Mean Product (thousand barrels) |
|--------|---------------------------------|
| **0** | 92.50 |
| **1** | 68.83 |
| **2** | 95.00 |

---

## Findings

- All regions have mean reserves below the break-even level of **111.11 thousand barrels**.  
- Profit is only achievable by selecting the **top-performing wells**, not the average well.  
- This makes prediction accuracy essential for profitability.

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>

Agreed. 

</div>

## Profit Function Explanation
The profit calculation function performs the following tasks:
- Sorts all predicted reserves in descending order
- Selects the top 200 wells
- Retrieves the actual reserves for these wells
- Computes profit using:
        
profit = (sum of selected actual reserves × 4500) − 100 000 000

# Single-Sample Profit Estimates (Before Bootstrapping)
Region	Profit (Single Sample)
0	33.21 million USD
1	24.15 million USD
2	27.10 million USD

Region 0 appears strongest in the single evaluation, but risk and variability must still be considered.


In [13]:
def bootstrap_profits(y_true, y_pred, n_bootstrap=1000, random_state=12345):
    state = np.random.RandomState(random_state)
    profits = []

    # Reset indices to avoid any alignment issues
    y_true = y_true.reset_index(drop=True)
    y_pred = pd.Series(y_pred).reset_index(drop=True)

    # We always simulate exploration of 500 wells
    SAMPLE_SIZE = 500

    for _ in range(n_bootstrap):
        # Sample 500 indices WITH replacement from the full set
        indices = state.randint(0, len(y_true), SAMPLE_SIZE)

        sample_true = y_true.iloc[indices]
        sample_pred = y_pred.iloc[indices]

        profit = calculate_profit(sample_true, sample_pred)
        profits.append(profit)

    return np.array(profits)

<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>


The bootstrapping looks great! We indeed need to sample 500 rows.
    
</div>

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    

However, we indeed need to sample 500 rows with replacement. The point is that if we remove the replacement, our subsample (or rather, all subsamples) will simply resemble the original sample. The question is, how do we ensure independence then? That is, there will be no variability in the data. As for duplicates, in reality, we can also encounter very similar wells. For simplicity, let's take this example: we have 100 objects and sample 100 as well. With replacement, we get different samples where some of the 100 elements may be missing, while others may appear multiple times. The example is very simple, but the idea should be clear. In our case, the original sample is much larger, and the subsample sizes are much smaller, which provides even greater variability in the data.   

</div>

In [14]:

def calculate_profit(y_true, y_pred):
    # Number of wells we are allowed to develop
    WELLS_TO_DEVELOP = 200
    
    # Development budget
    BUDGET = 100_000_000
    
    # Revenue per thousand barrels
    PRICE_PER_THOUSAND_BARRELS = 4500

    # 1. Select the TOP 200 wells by predicted product
    top_indices = y_pred.sort_values(ascending=False).head(WELLS_TO_DEVELOP).index

    # 2. Extract the TRUE product values for those wells
    selected_true = y_true.loc[top_indices]

    # 3. Total product (thousand barrels)
    total_product = selected_true.sum()

    # 4. Compute revenue and profit
    revenue = total_product * PRICE_PER_THOUSAND_BARRELS
    profit = revenue - BUDGET

    return profit

In [15]:
region_stats = []

for i, res in enumerate(results):
    y_true = res['y_valid']
    y_pred = res['y_pred']

    profits = bootstrap_profits(y_true, y_pred)

    mean_profit = profits.mean()
    conf_interval = np.percentile(profits, [2.5, 97.5])
    risk = (profits < 0).mean()

    print(f"\nRegion {i}")
    print(f"Average profit: {mean_profit/1_000_000:.2f} million USD")
    print(f"95% CI: {conf_interval[0]/1_000_000:.2f} to {conf_interval[1]/1_000_000:.2f} million USD")
    print(f"Risk of loss: {risk*100:.2f}%")


Region 0
Average profit: 6.01 million USD
95% CI: 0.13 to 12.31 million USD
Risk of loss: 2.00%

Region 1
Average profit: 6.65 million USD
95% CI: 1.58 to 11.98 million USD
Risk of loss: 0.30%

Region 2
Average profit: 6.16 million USD
95% CI: -0.12 to 12.31 million USD
Risk of loss: 3.00%


<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment 2 </b>
    
Correct. 
</div>

<div style="border: 5px solid red; padding: 10px; margin: 5px">
<b>   Reviewer's comment </b>
    

Hint: risk will not be 0 or 100. 

</div>

## Bootstrapping: Risk & Profit Analysis
Bootstrapping (1000 samples) simulates repeated exploration of 500 wells, allowing evaluation of:
- Average expected profit
- 95% confidence interval
- Probability of financial loss
    
Bootstrapping Results
    
Region 0
Average profit: 166.58 million USD
95% CI: 134.27 to 201.85 million USD
Risk of loss: 0.00%

Region 1
Average profit: 148.07 million USD
95% CI: 120.97 to 178.72 million USD
Risk of loss: 0.00%

Region 2
Average profit: 154.34 million USD
95% CI: 123.23 to 188.96 million USD
Risk of loss: 0.00%


In [16]:
valid_regions = [s for s in region_stats if s['risk'] < 0.025]

if valid_regions:
    best = max(valid_regions, key=lambda x: x['mean_profit'])
    print(f'\nRecommended region: {best["region"]}')
    print(f'Mean profit: {best["mean_profit"]/1_000_000:.2f} million USD')
    print(f'Risk of loss: {best["risk"]*100:.2f}%')
else:
    print('\nNo region satisfies the risk constraint of < 2.5%.')


No region satisfies the risk constraint of < 2.5%.


## Final Results & Recommendation
Summary of All Regions

After applying bootstrapping with 500-well samples and selecting the top 200 wells by predicted production, the risk and profit metrics were calculated for each region.

Region 0: Moderate profit (6.01M) and a 2.0% risk of loss.

Region 1: Highest profit (6.65M) and the lowest risk at only 0.30%.

Region 2: Similar average profit (6.16M), but highest risk (3.0%) and CI dips slightly below zero.

📌 Recommendation:

Region 1 should be selected — it offers the best balance of high profitability and minimum risk, making it the optimal region for oil well development.


<div style="border: 5px solid green; padding: 10px; margin: 5px">
<b>   Reviewer's comment 2 </b>
    
Excellent! 

</div>